### PR TITLE
interfaces/builtin/cpu-control: Add IRQ affinity control to cpu_control

### PR DESCRIPTION
**Pull request by locnnil**
Many real-time system need to tune IRQ affinity by accessing the related `/proc` file system files. 

Adding this to CPU-control avoids the need to create a new interface for IRQ's or either for real-time applications.

### CODE REVIEW

File: interfaces/builtin/cpu_control.go [hunk 1]
```diff
@@ -82,6 +82,11 @@ const cpuControlConnectedPlugAppArmor = `
 
 # Allow control CPU C-states switching see: https://docs.kernel.org/power/pm_qos_interface.html#pm-qos-framework
 /dev/cpu_dma_latency rw,
+
+# Allow interrupt affinity settings
```
Comment (bboozzoo) [inline interfaces/builtin/cpu_control.go:86]: could you also add a link to the kernel docs? https://www.kernel.org/doc/html/latest/core-api/irq/irq-affinity.html

Comment (locnnil) [reply interfaces/builtin/cpu_control.go]: Off course, sure!

**Review by bboozzoo — APPROVED** [review]
LGTM

File: interfaces/builtin/cpu_control.go [hunk 2]
```diff
@@ -82,6 +82,11 @@ const cpuControlConnectedPlugAppArmor = `
 
 # Allow control CPU C-states switching see: https://docs.kernel.org/power/pm_qos_interface.html#pm-qos-framework
 /dev/cpu_dma_latency rw,
+
+# Allow interrupt affinity settings, see https://www.kernel.org/doc/html/latest/core-api/irq/irq-affinity.html
+/proc/interrupts r,
+/proc/irq/[0-9]+/smp_affinity rw,
```
Comment (alexmurray) [inline interfaces/builtin/cpu_control.go:88]: should we also allow `/proc/irq/default_smp_affinity` ?

**Review by alexmurray — CHANGES_REQUESTED** [review]
Please use the `@{PROC}` variable rather than `/proc` and also I think perhaps we should also add `/proc/ieq/default_smp_affinity` too

**Comment by zyga** [comment]
@alexmurray as much as I understand abstractions, I don't get the value of `@{PROC}`. Do we expect proc to be mounted at another prefix? Do we expect `@{PROC}` to expand to something else in the future? Why do we have `@{PROC}` but not `@{SYS}` or `@{DEV}`.

**Review by zyga — APPROVED** [review]
The two listed files are the only writable files in the per-IRQ subdirectory so this looks good to me.

**Comment by alexmurray** [comment]
> @alexmurray as much as I understand abstractions, I don't get the value of `@{PROC}`. Do we expect proc to be mounted at another prefix? Do we expect `@{PROC}` to expand to something else in the future? Why do we have `@{PROC}` but not `@{SYS}` or `@{DEV}`.

This is a good question - fwiw there is actually a `@{sys}` variable (note the lower-case) which is defined as `/sys/` in `abstractions/kernelvars` but we don't seem to make as much use of it as `@{PROC}`. My understanding is that apparmor itself supports a configurable location of the procfs via this variable since it is possible it may not be mounted under `/proc` (however I suspect this would break many other things if it were not the case so perhaps this is just wishful thinking of a bygone era?)

I am not sure what the convention is around CAPS vs lowercase vars - @jrjohansen perhaps you may be able to provide any insights here.

But as far as snapd apparmor policy is concerned I think the most important this is consistency - and since we already use `@{PROC}` I think we should keep using it.

File: interfaces/builtin/cpu_control.go (hunk 2 above)
Comment (locnnil) [reply interfaces/builtin/cpu_control.go:88]: Yeah! It's a good idea! Thanks for the suggestion!

**Comment by locnnil** [comment]
@alexmurray, may I ask for your final decision on using `@{PROC}`? Should I replace `/proc` for `@{PROC}` or leave as it is?

File: interfaces/builtin/cpu_control.go (hunk 2 above)
Comment (locnnil) [reply interfaces/builtin/cpu_control.go:88]: Added on https://github.com/canonical/snapd/pull/14424/commits/8cce4658160770ae3677181707faa36d26587a51. Thank you!

**Comment by alexmurray** [comment]
Re the use of `@{PROC}` vs `/proc` - it is really about consistency - overall in snapd we have more uses of `@{PROC}` compared to `/proc`:

```
$ git grep '@{PROC}' interfaces/builtin/ | wc -l
256
$ git grep '/proc/' interfaces/builtin/ | wc -l
75
```

But as you say in this case (and sorry for not noticing this earlier) the cpu-control interface already uses `/proc` so in this case the most consistent implementation would be to either replace all uses of `/proc` with `@{PROC}` in cpu-control or just use `/proc` as you have done and leave the issue of consistency across snapd itself for another PR.

**Review by alexmurray — APPROVED** [review]
LGTM!

**Pull request merged**